restli-server/src/main/java/com/linkedin/restli/internal/server/response/BatchFinderResponseBuilder.java

```diff
@@ -207,14 +208,33 @@ private DataList getCriteriaParameters(RoutingResult routingResult)
     return(DataList)routingResult.getContext().getStructuredParameter(batchParamName);
   }
 
-  private URI buildCriteriaURI(ResourceContextImpl resourceContext, RecordTemplate criteria, String queryParam, URI uri)
+  private URI buildCriteriaURI(ResourceContextImpl resourceContext, RecordTemplate criteria, String batchParamName, URI uri)
   {
     DataList criteriaList = new DataList(1);
     criteriaList.add(criteria.data());
+    DataMap queryParams = extractQueryParamsFromResourceContext(resourceContext);
     return URIParamUtils.replaceQueryParam(uri,
-                                           queryParam,
+                                           batchParamName,
                                            criteriaList,
-                                           resourceContext.getParameters(),
+                                           queryParams,
                                            resourceContext.getRestliProtocolVersion());
   }
+
+  private DataMap extractQueryParamsFromResourceContext(ResourceContextImpl resourceContext)
+  {
+    DataMap queryParams = resourceContext.getParameters();
```

Comment:
You should create a copy instead of modifying the underlying parameters. There might be logic elsewhere (and custom filter) that depend on these parameters parsed from URI.

Reply:
Got it, DataMap has two copy related method, clone and copy, the clone seems to be a shallow copy and the copy seems to be suit for this use case